Stylecop reports:
```
OpenRA.Mods.Common/Widgets/Logic/Editor/TileSelectorLogic.cs(91,8): error CS0219: Warning as Error: The variable `categorySelector' is assigned but its value is never used
```

--- a/OpenRA.Mods.Common/OpenRA.Mods.Common.csproj
+++ b/OpenRA.Mods.Common/OpenRA.Mods.Common.csproj
@@ -694,6 +694,7 @@
     <Compile Include="Widgets\Logic\Lobby\SpawnSelectorTooltipLogic.cs" />
     <Compile Include="Widgets\Logic\MainMenuLogic.cs" />
     <Compile Include="Widgets\Logic\Editor\ActorSelectorLogic.cs" />
+    <Compile Include="Widgets\Logic\Editor\CommonSelectorLogic.cs" />
     <Compile Include="Widgets\Logic\Editor\LayerSelectorLogic.cs" />
     <Compile Include="Widgets\Logic\Editor\MapEditorLogic.cs" />
     <Compile Include="Widgets\Logic\Editor\MapEditorTabsLogic.cs" />

--- a/OpenRA.Mods.Common/Widgets/Logic/Editor/ActorSelectorLogic.cs
+++ b/OpenRA.Mods.Common/Widgets/Logic/Editor/ActorSelectorLogic.cs
@@ -19,7 +19,7 @@ using OpenRA.Widgets;
 
 namespace OpenRA.Mods.Common.Widgets.Logic
 {
-	public class ActorSelectorLogic : ChromeLogic
+	public class ActorSelectorLogic : CommonSelectorLogic
 	{
 		class ActorSelectorActor
 		{
@@ -37,30 +37,21 @@ namespace OpenRA.Mods.Common.Widgets.Logic
 			}
 		}
 
-		readonly EditorViewportControllerWidget editor;
 		readonly DropDownButtonWidget ownersDropDown;
 		readonly ScrollPanelWidget panel;
 		readonly ScrollItemWidget itemTemplate;
 		readonly Ruleset mapRules;
-		readonly World world;
-		readonly WorldRenderer worldRenderer;
-		readonly string[] allCategories;
-		readonly HashSet<string> selectedCategories = new HashSet<string>();
-		readonly List<string> filteredCategories = new List<string>();
 
 		readonly ActorSelectorActor[] allActors;
 
 		PlayerReference selectedOwner;
-		string searchFilter;
 
 		[ObjectCreator.UseCtor]
 		public ActorSelectorLogic(Widget widget, World world, WorldRenderer worldRenderer)
+			: base(widget, world, worldRenderer)
 		{
 			mapRules = world.Map.Rules;
-			this.world = world;
-			this.worldRenderer = worldRenderer;
 
-			editor = widget.Parent.Get<EditorViewportControllerWidget>("MAP_EDITOR");
 			ownersDropDown = widget.Get<DropDownButtonWidget>("OWNERS_DROPDOWN");
 
 			panel = widget.Get<ScrollPanelWidget>("ACTORTEMPLATE_LIST");
@@ -130,7 +121,7 @@ namespace OpenRA.Mods.Common.Widgets.Logic
 
 				var tooltipText = (tooltip == null ? "Type: " : tooltip.Name + "\nType: ") + a.Name;
 				allActorsTemp.Add(new ActorSelectorActor(a, editorData.Categories, searchTerms.ToArray(), tooltipText));
- 			}
+			}
 
 			allActors = allActorsTemp.ToArray();
 
@@ -141,27 +132,27 @@ namespace OpenRA.Mods.Common.Widgets.Logic
 
 			foreach (var c in allCategories)
 			{
-				selectedCategories.Add(c);
-				filteredCategories.Add(c);
+				SelectedCategories.Add(c);
+				FilteredCategories.Add(c);
 			}
 
 			var searchTextField = widget.Get<TextFieldWidget>("SEARCH_TEXTFIELD");
 			searchTextField.OnTextEdited = () =>
 			{
 				searchFilter = searchTextField.Text.Trim();
-				filteredCategories.Clear();
+				FilteredCategories.Clear();
 
 				if (!string.IsNullOrEmpty(searchFilter))
-					filteredCategories.AddRange(
+					FilteredCategories.AddRange(
 						allActors.Where(t => t.SearchTerms.Any(
 							s => s.IndexOf(searchFilter, StringComparison.OrdinalIgnoreCase) >= 0))
 						.SelectMany(t => t.Categories)
 						.Distinct()
 						.OrderBy(x => x));
 				else
-					filteredCategories.AddRange(allCategories);
+					FilteredCategories.AddRange(allCategories);
 
-				InitializeActorPreviews();
+				InitializePreviews();
 			};
 
 			searchTextField.OnEscKey = () =>
@@ -174,16 +165,16 @@ namespace OpenRA.Mods.Common.Widgets.Logic
 			var actorCategorySelector = widget.Get<DropDownButtonWidget>("CATEGORIES_DROPDOWN");
 			actorCategorySelector.GetText = () =>
 			{
-				if (selectedCategories.Count == 0)
+				if (SelectedCategories.Count == 0)
 					return "None";
 
 				if (!string.IsNullOrEmpty(searchFilter))
 					return "Search Results";
 
-				if (selectedCategories.Count == 1)
-					return selectedCategories.First();
+				if (SelectedCategories.Count == 1)
+					return SelectedCategories.First();
 
-				if (selectedCategories.Count == allCategories.Length)
+				if (SelectedCategories.Count == allCategories.Length)
 					return "All";
 
 				return "Multiple";
@@ -195,10 +186,10 @@ namespace OpenRA.Mods.Common.Widgets.Logic
 					searchTextField.YieldKeyboardFocus();
 
 				actorCategorySelector.RemovePanel();
-				actorCategorySelector.AttachPanel(CreateCategoriesPanel());
+				actorCategorySelector.AttachPanel(CreateCategoriesPanel(panel));
 			};
 
-			InitializeActorPreviews();
+			InitializePreviews();
 		}
 
 		void SelectOwner(PlayerReference option)
@@ -206,67 +197,18 @@ namespace OpenRA.Mods.Common.Widgets.Logic
 			selectedOwner = option;
 			ownersDropDown.Text = option.Name;
 			ownersDropDown.TextColor = option.Color.RGB;
-			InitializeActorPreviews();
+			InitializePreviews();
 		}
 
-		Widget CreateCategoriesPanel()
-		{
-			var categoriesPanel = Ui.LoadWidget("ACTOR_CATEGORY_FILTER_PANEL", null, new WidgetArgs());
-			var categoryTemplate = categoriesPanel.Get<CheckboxWidget>("CATEGORY_TEMPLATE");
-
-			var selectButtons = categoriesPanel.Get<ContainerWidget>("SELECT_CATEGORIES_BUTTONS");
-			categoriesPanel.AddChild(selectButtons);
-
-			var selectAll = selectButtons.Get<ButtonWidget>("SELECT_ALL");
-			selectAll.OnClick = () =>
-			{
-				selectedCategories.Clear();
-				foreach (var c in allCategories)
-					selectedCategories.Add(c);
-
-				InitializeActorPreviews();
-			};
-
-			var selectNone = selectButtons.Get<ButtonWidget>("SELECT_NONE");
-			selectNone.OnClick = () =>
-			{
-				selectedCategories.Clear();
-				InitializeActorPreviews();
-			};
-
-			var categoryHeight = 5 + selectButtons.Bounds.Height;
-			foreach (var cat in filteredCategories)
-			{
-				var category = (CheckboxWidget)categoryTemplate.Clone();
-				category.GetText = () => cat;
-				category.IsChecked = () => selectedCategories.Contains(cat);
-				category.IsVisible = () => true;
-				category.OnClick = () =>
-				{
-					if (!selectedCategories.Remove(cat))
-						selectedCategories.Add(cat);
-
-					InitializeActorPreviews();
-				};
-
-				categoriesPanel.AddChild(category);
-				categoryHeight += categoryTemplate.Bounds.Height;
-			}
-
-			categoriesPanel.Bounds.Height = Math.Min(categoryHeight, panel.Bounds.Height);
-
-			return categoriesPanel;
-		}
-
-		void InitializeActorPreviews()
+		protected override void InitializePreviews()
 		{
 			panel.RemoveChildren();
-			if (!selectedCategories.Any())
+			if (!SelectedCategories.Any())
 				return;
 
 			foreach (var a in allActors)
 			{
-				if (!selectedCategories.Overlaps(a.Categories))
+				if (!SelectedCategories.Overlaps(a.Categories))
 					continue;
 
 				if (!string.IsNullOrEmpty(searchFilter) && !a.SearchTerms.Any(s => s.IndexOf(searchFilter, StringComparison.OrdinalIgnoreCase) >= 0))
@@ -283,8 +225,8 @@ namespace OpenRA.Mods.Common.Widgets.Logic
 				try
 				{
 					var item = ScrollItemWidget.Setup(itemTemplate,
-						() => { var brush = editor.CurrentBrush as EditorActorBrush; return brush != null && brush.Actor == actor; },
-						() => editor.SetBrush(new EditorActorBrush(editor, actor, selectedOwner, worldRenderer)));
+						() => { var brush = Editor.CurrentBrush as EditorActorBrush; return brush != null && brush.Actor == actor; },
+						() => Editor.SetBrush(new EditorActorBrush(Editor, actor, selectedOwner, WorldRenderer)));
 
 					var preview = item.Get<ActorPreviewWidget>("ACTOR_PREVIEW");
 					preview.SetPreview(actor, td);
@@ -309,7 +251,7 @@ namespace OpenRA.Mods.Common.Widgets.Logic
 				catch
 				{
 					Log.Write("debug", "Map editor ignoring actor {0}, because of missing sprites for tileset {1}.",
-						actor.Name, world.Map.Rules.TileSet.Id);
+						actor.Name, World.Map.Rules.TileSet.Id);
 					continue;
 				}
 			}

--- a/OpenRA.Mods.Common/Widgets/Logic/Editor/ActorSelectorLogic.cs
+++ b/OpenRA.Mods.Common/Widgets/Logic/Editor/ActorSelectorLogic.cs
@@ -38,26 +38,17 @@ namespace OpenRA.Mods.Common.Widgets.Logic
 		}
 
 		readonly DropDownButtonWidget ownersDropDown;
-		readonly ScrollPanelWidget panel;
-		readonly ScrollItemWidget itemTemplate;
 		readonly Ruleset mapRules;
-
 		readonly ActorSelectorActor[] allActors;
 
 		PlayerReference selectedOwner;
 
 		[ObjectCreator.UseCtor]
 		public ActorSelectorLogic(Widget widget, World world, WorldRenderer worldRenderer)
-			: base(widget, world, worldRenderer)
+			: base(widget, world, worldRenderer, "ACTORTEMPLATE_LIST", "ACTORPREVIEW_TEMPLATE")
 		{
 			mapRules = world.Map.Rules;
-
 			ownersDropDown = widget.Get<DropDownButtonWidget>("OWNERS_DROPDOWN");
-
-			panel = widget.Get<ScrollPanelWidget>("ACTORTEMPLATE_LIST");
-			itemTemplate = panel.Get<ScrollItemWidget>("ACTORPREVIEW_TEMPLATE");
-			panel.Layout = new GridLayout(panel);
-
 			var editorLayer = world.WorldActor.Trait<EditorActorLayer>();
 
 			selectedOwner = editorLayer.Players.Players.Values.First();
@@ -136,10 +127,9 @@ namespace OpenRA.Mods.Common.Widgets.Logic
 				FilteredCategories.Add(c);
 			}
 
-			var searchTextField = widget.Get<TextFieldWidget>("SEARCH_TEXTFIELD");
-			searchTextField.OnTextEdited = () =>
+			SearchTextField.OnTextEdited = () =>
 			{
-				searchFilter = searchTextField.Text.Trim();
+				searchFilter = SearchTextField.Text.Trim();
 				FilteredCategories.Clear();
 
 				if (!string.IsNullOrEmpty(searchFilter))
@@ -155,40 +145,6 @@ namespace OpenRA.Mods.Common.Widgets.Logic
 				InitializePreviews();
 			};
 
-			searchTextField.OnEscKey = () =>
-			{
-				searchTextField.Text = "";
-				searchTextField.YieldKeyboardFocus();
-				return true;
-			};
-
-			var actorCategorySelector = widget.Get<DropDownButtonWidget>("CATEGORIES_DROPDOWN");
-			actorCategorySelector.GetText = () =>
-			{
-				if (SelectedCategories.Count == 0)
-					return "None";
-
-				if (!string.IsNullOrEmpty(searchFilter))
-					return "Search Results";
-
-				if (SelectedCategories.Count == 1)
-					return SelectedCategories.First();
-
-				if (SelectedCategories.Count == allCategories.Length)
-					return "All";
-
-				return "Multiple";
-			};
-
-			actorCategorySelector.OnMouseDown = _ =>
-			{
-				if (searchTextField != null)
-					searchTextField.YieldKeyboardFocus();
-
-				actorCategorySelector.RemovePanel();
-				actorCategorySelector.AttachPanel(CreateCategoriesPanel(panel));
-			};
-
 			InitializePreviews();
 		}
 
@@ -202,7 +158,7 @@ namespace OpenRA.Mods.Common.Widgets.Logic
 
 		protected override void InitializePreviews()
 		{
-			panel.RemoveChildren();
+			Panel.RemoveChildren();
 			if (!SelectedCategories.Any())
 				return;
 
@@ -224,7 +180,7 @@ namespace OpenRA.Mods.Common.Widgets.Logic
 
 				try
 				{
-					var item = ScrollItemWidget.Setup(itemTemplate,
+					var item = ScrollItemWidget.Setup(ItemTemplate,
 						() => { var brush = Editor.CurrentBrush as EditorActorBrush; return brush != null && brush.Actor == actor; },
 						() => Editor.SetBrush(new EditorActorBrush(Editor, actor, selectedOwner, WorldRenderer)));
 
@@ -233,8 +189,8 @@ namespace OpenRA.Mods.Common.Widgets.Logic
 
 					// Scale templates to fit within the panel
 					var scale = 1f;
-					if (scale * preview.IdealPreviewSize.X > itemTemplate.Bounds.Width)
-						scale = (itemTemplate.Bounds.Width - panel.ItemSpacing) / (float)preview.IdealPreviewSize.X;
+					if (scale * preview.IdealPreviewSize.X > ItemTemplate.Bounds.Width)
+						scale = (ItemTemplate.Bounds.Width - Panel.ItemSpacing) / (float)preview.IdealPreviewSize.X;
 
 					preview.GetScale = () => scale;
 					preview.Bounds.Width = (int)(scale * preview.IdealPreviewSize.X);
@@ -246,7 +202,7 @@ namespace OpenRA.Mods.Common.Widgets.Logic
 
 					item.GetTooltipText = () => a.Tooltip;
 
-					panel.AddChild(item);
+					Panel.AddChild(item);
 				}
 				catch
 				{

--- a/OpenRA.Mods.Common/Widgets/Logic/Editor/CommonSelectorLogic.cs
+++ b/OpenRA.Mods.Common/Widgets/Logic/Editor/CommonSelectorLogic.cs
@@ -1,0 +1,91 @@
+﻿#region Copyright & License Information
+/*
+ * Copyright 2007-2017 The OpenRA Developers (see AUTHORS)
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation, either version 3 of
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+
+using System;
+using System.Collections.Generic;
+using OpenRA.Graphics;
+using OpenRA.Widgets;
+
+namespace OpenRA.Mods.Common.Widgets.Logic
+{
+	public abstract class CommonSelectorLogic : ChromeLogic
+	{
+		protected readonly Widget Widget;
+		protected readonly World World;
+		protected readonly WorldRenderer WorldRenderer;
+		protected readonly EditorViewportControllerWidget Editor;
+
+		protected readonly HashSet<string> SelectedCategories = new HashSet<string>();
+		protected readonly List<string> FilteredCategories = new List<string>();
+
+		protected string[] allCategories;
+		protected string searchFilter;
+
+		public CommonSelectorLogic(Widget widget, World world, WorldRenderer worldRenderer)
+		{
+			this.Widget = widget;
+			this.World = world;
+			this.WorldRenderer = worldRenderer;
+			Editor = widget.Parent.Get<EditorViewportControllerWidget>("MAP_EDITOR");
+		}
+
+		protected Widget CreateCategoriesPanel(ScrollPanelWidget panel)
+		{
+			var categoriesPanel = Ui.LoadWidget("ACTOR_CATEGORY_FILTER_PANEL", null, new WidgetArgs());
+			var categoryTemplate = categoriesPanel.Get<CheckboxWidget>("CATEGORY_TEMPLATE");
+
+			var selectButtons = categoriesPanel.Get<ContainerWidget>("SELECT_CATEGORIES_BUTTONS");
+			categoriesPanel.AddChild(selectButtons);
+
+			var selectAll = selectButtons.Get<ButtonWidget>("SELECT_ALL");
+			selectAll.OnClick = () =>
+			{
+				SelectedCategories.Clear();
+				foreach (var c in allCategories)
+					SelectedCategories.Add(c);
+
+				InitializePreviews();
+			};
+
+			var selectNone = selectButtons.Get<ButtonWidget>("SELECT_NONE");
+			selectNone.OnClick = () =>
+			{
+				SelectedCategories.Clear();
+				InitializePreviews();
+			};
+
+			var categoryHeight = 5 + selectButtons.Bounds.Height;
+			foreach (var cat in FilteredCategories)
+			{
+				var category = (CheckboxWidget)categoryTemplate.Clone();
+				category.GetText = () => cat;
+				category.IsChecked = () => SelectedCategories.Contains(cat);
+				category.IsVisible = () => true;
+				category.OnClick = () =>
+				{
+					if (!SelectedCategories.Remove(cat))
+						SelectedCategories.Add(cat);
+
+					InitializePreviews();
+				};
+
+				categoriesPanel.AddChild(category);
+				categoryHeight += categoryTemplate.Bounds.Height;
+			}
+
+			categoriesPanel.Bounds.Height = Math.Min(categoryHeight, panel.Bounds.Height);
+
+			return categoriesPanel;
+		}
+
+		protected abstract void InitializePreviews();
+	}
+}

--- a/OpenRA.Mods.Common/Widgets/Logic/Editor/CommonSelectorLogic.cs
+++ b/OpenRA.Mods.Common/Widgets/Logic/Editor/CommonSelectorLogic.cs
@@ -11,6 +11,7 @@
 
 using System;
 using System.Collections.Generic;
+using System.Linq;
 using OpenRA.Graphics;
 using OpenRA.Widgets;
 
@@ -19,9 +20,12 @@ namespace OpenRA.Mods.Common.Widgets.Logic
 	public abstract class CommonSelectorLogic : ChromeLogic
 	{
 		protected readonly Widget Widget;
+		protected readonly TextFieldWidget SearchTextField;
 		protected readonly World World;
 		protected readonly WorldRenderer WorldRenderer;
 		protected readonly EditorViewportControllerWidget Editor;
+		protected readonly ScrollPanelWidget Panel;
+		protected readonly ScrollItemWidget ItemTemplate;
 
 		protected readonly HashSet<string> SelectedCategories = new HashSet<string>();
 		protected readonly List<string> FilteredCategories = new List<string>();
@@ -29,17 +33,55 @@ namespace OpenRA.Mods.Common.Widgets.Logic
 		protected string[] allCategories;
 		protected string searchFilter;
 
-		public CommonSelectorLogic(Widget widget, World world, WorldRenderer worldRenderer)
+		public CommonSelectorLogic(Widget widget, World world, WorldRenderer worldRenderer, string templateListId, string previewTemplateId)
 		{
 			this.Widget = widget;
 			this.World = world;
 			this.WorldRenderer = worldRenderer;
 			Editor = widget.Parent.Get<EditorViewportControllerWidget>("MAP_EDITOR");
+			Panel = widget.Get<ScrollPanelWidget>(templateListId);
+			ItemTemplate = Panel.Get<ScrollItemWidget>(previewTemplateId);
+			Panel.Layout = new GridLayout(Panel);
+
+			SearchTextField = widget.Get<TextFieldWidget>("SEARCH_TEXTFIELD");
+			SearchTextField.OnEscKey = () =>
+			{
+				SearchTextField.Text = "";
+				SearchTextField.YieldKeyboardFocus();
+				return true;
+			};
+
+			var categorySelector = widget.Get<DropDownButtonWidget>("CATEGORIES_DROPDOWN");
+			categorySelector.GetText = () =>
+			{
+				if (SelectedCategories.Count == 0)
+					return "None";
+
+				if (!string.IsNullOrEmpty(searchFilter))
+					return "Search Results";
+
+				if (SelectedCategories.Count == 1)
+					return SelectedCategories.First();
+
+				if (SelectedCategories.Count == allCategories.Length)
+					return "All";
+
+				return "Multiple";
+			};
+
+			categorySelector.OnMouseDown = _ =>
+			{
+				if (SearchTextField != null)
+					SearchTextField.YieldKeyboardFocus();
+
+				categorySelector.RemovePanel();
+				categorySelector.AttachPanel(CreateCategoriesPanel(Panel));
+			};
 		}
 
 		protected Widget CreateCategoriesPanel(ScrollPanelWidget panel)
 		{
-			var categoriesPanel = Ui.LoadWidget("ACTOR_CATEGORY_FILTER_PANEL", null, new WidgetArgs());
+			var categoriesPanel = Ui.LoadWidget("CATEGORY_FILTER_PANEL", null, new WidgetArgs());
 			var categoryTemplate = categoriesPanel.Get<CheckboxWidget>("CATEGORY_TEMPLATE");
 
 			var selectButtons = categoriesPanel.Get<ContainerWidget>("SELECT_CATEGORIES_BUTTONS");

--- a/OpenRA.Mods.Common/Widgets/Logic/Editor/TileSelectorLogic.cs
+++ b/OpenRA.Mods.Common/Widgets/Logic/Editor/TileSelectorLogic.cs
@@ -35,87 +35,44 @@ namespace OpenRA.Mods.Common.Widgets.Logic
 		}
 
 		readonly TileSet tileset;
-		readonly ScrollPanelWidget panel;
-		readonly ScrollItemWidget itemTemplate;
 		readonly TileSelectorTemplate[] allTemplates;
 
-		string selectedCategory;
-		string userSelectedCategory;
-
 		[ObjectCreator.UseCtor]
-		public TileSelectorLogic(Widget widget, World world, WorldRenderer worldRenderer) :
-			base(widget, world, worldRenderer)
+		public TileSelectorLogic(Widget widget, World world, WorldRenderer worldRenderer)
+			: base(widget, world, worldRenderer, "TILETEMPLATE_LIST", "TILEPREVIEW_TEMPLATE")
 		{
 			tileset = world.Map.Rules.TileSet;
-
-			panel = widget.Get<ScrollPanelWidget>("TILETEMPLATE_LIST");
-			itemTemplate = panel.Get<ScrollItemWidget>("TILEPREVIEW_TEMPLATE");
-			panel.Layout = new GridLayout(panel);
-
 			allTemplates = tileset.Templates.Values.Select(t => new TileSelectorTemplate(t)).ToArray();
 
-			var orderedCategories = allTemplates.SelectMany(t => t.Categories)
+			allCategories = allTemplates.SelectMany(t => t.Categories)
 				.Distinct()
 				.OrderBy(CategoryOrder)
 				.ToArray();
 
-			var searchTextField = widget.Get<TextFieldWidget>("SEARCH_TEXTFIELD");
-			searchTextField.OnTextEdited = () =>
+			foreach (var c in allCategories)
 			{
-				searchFilter = searchTextField.Text.Trim();
-				selectedCategory = string.IsNullOrEmpty(searchFilter) ? userSelectedCategory : null;
+				SelectedCategories.Add(c);
+				FilteredCategories.Add(c);
+			}
+
+			SearchTextField.OnTextEdited = () =>
+			{
+				searchFilter = SearchTextField.Text.Trim();
+				FilteredCategories.Clear();
+
+				if (!string.IsNullOrEmpty(searchFilter))
+					FilteredCategories.AddRange(
+						allTemplates.Where(t => t.SearchTerms.Any(
+							s => s.IndexOf(searchFilter, StringComparison.OrdinalIgnoreCase) >= 0))
+						.SelectMany(t => t.Categories)
+						.Distinct()
+						.OrderBy(CategoryOrder));
+				else
+					FilteredCategories.AddRange(allCategories);
 
 				InitializePreviews();
 			};
 
-			searchTextField.OnEscKey = () =>
-			{
-				searchTextField.Text = "";
-				searchTextField.YieldKeyboardFocus();
-				return true;
-			};
-
-			Func<string, string> categoryTitle = s => s != null ? s : "Search Results";
-			Func<string, ScrollItemWidget, ScrollItemWidget> setupItem = (option, template) =>
-			{
-				var item = ScrollItemWidget.Setup(template,	() => selectedCategory == option, () =>
-				{
-					selectedCategory = option;
-					if (option != null)
-						userSelectedCategory = option;
-
-					InitializePreviews();
-				});
-
-				var title = categoryTitle(option);
-				item.Get<LabelWidget>("LABEL").GetText = () => title;
-				return item;
-			};
-
-			var tileCategorySelector = widget.Get<DropDownButtonWidget>("CATEGORIES_DROPDOWN");
-			tileCategorySelector.OnClick = () =>
-			{
-				if (searchTextField != null)
-					searchTextField.YieldKeyboardFocus();
-
-				var categories = orderedCategories.AsEnumerable();
-				if (!string.IsNullOrEmpty(searchFilter))
-				{
-					var filteredCategories = allTemplates.Where(t => t.SearchTerms.Any(
-							s => s.IndexOf(searchFilter, StringComparison.OrdinalIgnoreCase) >= 0))
-						.SelectMany(t => t.Categories)
-						.Distinct()
-						.OrderBy(CategoryOrder);
-					categories = new string[] { null }.Concat(filteredCategories);
-				}
-
-				tileCategorySelector.ShowDropDown("LABEL_DROPDOWN_TEMPLATE", 270, categories, setupItem);
-			};
-
-			var actorCategorySelector = widget.Get<DropDownButtonWidget>("CATEGORIES_DROPDOWN");
-			actorCategorySelector.GetText = () => categoryTitle(selectedCategory);
-
-			selectedCategory = userSelectedCategory = orderedCategories.First();
 			InitializePreviews();
 		}
 
@@ -127,18 +84,20 @@ namespace OpenRA.Mods.Common.Widgets.Logic
 
 		protected override void InitializePreviews()
 		{
-			panel.RemoveChildren();
+			Panel.RemoveChildren();
+			if (!SelectedCategories.Any())
+				return;
 
 			foreach (var t in allTemplates)
 			{
-				if (selectedCategory != null && !t.Categories.Contains(selectedCategory))
+				if (!SelectedCategories.Overlaps(t.Categories))
 					continue;
 
 				if (!string.IsNullOrEmpty(searchFilter) && !t.SearchTerms.Any(s => s.IndexOf(searchFilter, StringComparison.OrdinalIgnoreCase) >= 0))
 					continue;
 
 				var tileId = t.Template.Id;
-				var item = ScrollItemWidget.Setup(itemTemplate,
+				var item = ScrollItemWidget.Setup(ItemTemplate,
 					() => { var brush = Editor.CurrentBrush as EditorTileBrush; return brush != null && brush.Template == tileId; },
 					() => Editor.SetBrush(new EditorTileBrush(Editor, tileId, WorldRenderer)));
 
@@ -149,7 +108,7 @@ namespace OpenRA.Mods.Common.Widgets.Logic
 
 				// Scale templates to fit within the panel
 				var scale = 1f;
-				while (scale * bounds.Width > itemTemplate.Bounds.Width)
+				while (scale * bounds.Width > ItemTemplate.Bounds.Width)
 					scale /= 2;
 
 				preview.Template = template;
@@ -162,7 +121,7 @@ namespace OpenRA.Mods.Common.Widgets.Logic
 				item.IsVisible = () => true;
 				item.GetTooltipText = () => t.Tooltip;
 
-				panel.AddChild(item);
+				Panel.AddChild(item);
 			}
 		}
 	}

--- a/OpenRA.Mods.Common/Widgets/Logic/Editor/TileSelectorLogic.cs
+++ b/OpenRA.Mods.Common/Widgets/Logic/Editor/TileSelectorLogic.cs
@@ -16,7 +16,7 @@ using OpenRA.Widgets;
 
 namespace OpenRA.Mods.Common.Widgets.Logic
 {
-	public class TileSelectorLogic : ChromeLogic
+	public class TileSelectorLogic : CommonSelectorLogic
 	{
 		class TileSelectorTemplate
 		{
@@ -35,23 +35,19 @@ namespace OpenRA.Mods.Common.Widgets.Logic
 		}
 
 		readonly TileSet tileset;
-		readonly WorldRenderer worldRenderer;
-		readonly EditorViewportControllerWidget editor;
 		readonly ScrollPanelWidget panel;
 		readonly ScrollItemWidget itemTemplate;
 		readonly TileSelectorTemplate[] allTemplates;
 
 		string selectedCategory;
 		string userSelectedCategory;
-		string searchFilter;
 
 		[ObjectCreator.UseCtor]
-		public TileSelectorLogic(Widget widget, WorldRenderer worldRenderer)
+		public TileSelectorLogic(Widget widget, World world, WorldRenderer worldRenderer) :
+			base(widget, world, worldRenderer)
 		{
-			tileset = worldRenderer.World.Map.Rules.TileSet;
-			this.worldRenderer = worldRenderer;
+			tileset = world.Map.Rules.TileSet;
 
-			editor = widget.Parent.Get<EditorViewportControllerWidget>("MAP_EDITOR");
 			panel = widget.Get<ScrollPanelWidget>("TILETEMPLATE_LIST");
 			itemTemplate = panel.Get<ScrollItemWidget>("TILEPREVIEW_TEMPLATE");
 			panel.Layout = new GridLayout(panel);
@@ -69,7 +65,7 @@ namespace OpenRA.Mods.Common.Widgets.Logic
 				searchFilter = searchTextField.Text.Trim();
 				selectedCategory = string.IsNullOrEmpty(searchFilter) ? userSelectedCategory : null;
 
-				InitializeTilePreview();
+				InitializePreviews();
 			};
 
 			searchTextField.OnEscKey = () =>
@@ -88,7 +84,7 @@ namespace OpenRA.Mods.Common.Widgets.Logic
 					if (option != null)
 						userSelectedCategory = option;
 
-					InitializeTilePreview();
+					InitializePreviews();
 				});
 
 				var title = categoryTitle(option);
@@ -120,7 +116,7 @@ namespace OpenRA.Mods.Common.Widgets.Logic
 			actorCategorySelector.GetText = () => categoryTitle(selectedCategory);
 
 			selectedCategory = userSelectedCategory = orderedCategories.First();
-			InitializeTilePreview();
+			InitializePreviews();
 		}
 
 		int CategoryOrder(string category)
@@ -129,7 +125,7 @@ namespace OpenRA.Mods.Common.Widgets.Logic
 			return i >= 0 ? i : int.MaxValue;
 		}
 
-		void InitializeTilePreview()
+		protected override void InitializePreviews()
 		{
 			panel.RemoveChildren();
 
@@ -143,13 +139,13 @@ namespace OpenRA.Mods.Common.Widgets.Logic
 
 				var tileId = t.Template.Id;
 				var item = ScrollItemWidget.Setup(itemTemplate,
-					() => { var brush = editor.CurrentBrush as EditorTileBrush; return brush != null && brush.Template == tileId; },
-					() => editor.SetBrush(new EditorTileBrush(editor, tileId, worldRenderer)));
+					() => { var brush = Editor.CurrentBrush as EditorTileBrush; return brush != null && brush.Template == tileId; },
+					() => Editor.SetBrush(new EditorTileBrush(Editor, tileId, WorldRenderer)));
 
 				var preview = item.Get<TerrainTemplatePreviewWidget>("TILE_PREVIEW");
 				var template = tileset.Templates[tileId];
-				var grid = worldRenderer.World.Map.Grid;
-				var bounds = worldRenderer.Theater.TemplateBounds(template, grid.TileSize, grid.Type);
+				var grid = WorldRenderer.World.Map.Grid;
+				var bounds = WorldRenderer.Theater.TemplateBounds(template, grid.TileSize, grid.Type);
 
 				// Scale templates to fit within the panel
 				var scale = 1f;

--- a/mods/cnc/chrome/editor.yaml
+++ b/mods/cnc/chrome/editor.yaml
@@ -569,7 +569,7 @@ Container@EDITOR_WORLD_ROOT:
 			Font: Bold
 			Contrast: true
 
-ScrollPanel@ACTOR_CATEGORY_FILTER_PANEL:
+ScrollPanel@CATEGORY_FILTER_PANEL:
 	Width: 190
 	Children:
 		Container@SELECT_CATEGORIES_BUTTONS:

--- a/mods/common/chrome/editor.yaml
+++ b/mods/common/chrome/editor.yaml
@@ -548,7 +548,7 @@ Container@EDITOR_WORLD_ROOT:
 			Font: Bold
 			Contrast: true
 
-ScrollPanel@ACTOR_CATEGORY_FILTER_PANEL:
+ScrollPanel@CATEGORY_FILTER_PANEL:
 	Width: 170
 	Children:
 		Container@SELECT_CATEGORIES_BUTTONS:

